src/Processors/Formats/Impl/ConfluentRegistry.cpp

```diff
@@ -0,0 +1,170 @@
+    const std::string & base_url_, const String & logger_name_, [[maybe_unused]] size_t schema_cache_max_size)
+    : base_url(base_url_)
+#if USE_AVRO
+    , avro_schema_cache(CurrentMetrics::AvroSchemaCacheBytes, CurrentMetrics::AvroSchemaCacheCells, schema_cache_max_size)
+#endif
+#if USE_PROTOBUF
+    , protobuf_schema_cache(CurrentMetrics::ProtobufSchemaCacheBytes, CurrentMetrics::ProtobufSchemaCacheCells, schema_cache_max_size)
+#endif
+    , logger_name(logger_name_)
+{
+    if (base_url.empty())
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Empty Schema Registry URL");
+}
+
+#if USE_AVRO
+avro::ValidSchema ConfluentSchemaRegistry::getAvroSchema(uint32_t id)
+{
+    auto [result_schema, loaded] = avro_schema_cache.getOrSet(
+        id,
+        [this, id]()
+        {
+            try
+            {
+                auto avro_schema = fetchSchema(id);
+                return std::make_shared<avro::ValidSchema>(avro::compileJsonSchemaFromString(avro_schema));
+            }
+            catch (const avro::Exception & e)
+            {
+                throw Exception::createDeprecated(e.what(), ErrorCodes::INCORRECT_DATA);
+            }
+        });
+    return *result_schema;
+}
+#endif
+
+#if USE_PROTOBUF
+
+void ConfluentSchemaRegistry::ErrorCollector::RecordError(absl::string_view filename, int line, int column, absl::string_view message)
+{
+    throw Exception(ErrorCodes::INCORRECT_DATA, "Proto parse error at {} {}: {} {}", filename, line, column, message);
+}
+
+const google::protobuf::Descriptor * ConfluentSchemaRegistry::getProtobufSchema(uint32_t id)
+{
+    auto [schema, loaded] = protobuf_schema_cache.getOrSet(
+        id,
+        [this, id]()
+        {
+            auto schema_representation = fetchSchema(id);
+
+            ErrorCollector error_collector;
+            google::protobuf::compiler::DiskSourceTree source_tree;
+            source_tree.MapPath("", ".");
+
+            google::protobuf::compiler::SourceTreeDescriptorDatabase db(&source_tree);
+            google::protobuf::compiler::Importer importer(&source_tree, &error_collector);
+
+            std::istringstream // STYLE_CHECK_ALLOW_STD_STRING_STREAM
+                iss(schema_representation);
+            google::protobuf::io::IstreamInputStream input_stream(&iss);
+            google::protobuf::io::Tokenizer tokenizer(&input_stream, nullptr);
+            google::protobuf::compiler::Parser parser;
+
+            google::protobuf::FileDescriptorProto file_proto;
+
+            if (!parser.Parse(&tokenizer, &file_proto))
+                throw Exception(ErrorCodes::INCORRECT_DATA, "Failed to parse proto text");
+
+            file_proto.set_name(generator.createRandom().toString() + ".proto");
+            google::protobuf::FileDescriptorSet fds;
+            *fds.add_file() = file_proto;
+
+            return std::make_shared<const google::protobuf::Descriptor *>(independent_pool.BuildFile(file_proto)->message_type(0));
```

Comment:
If we return a pointer to `google::protobuf::Descriptor` then what it holds? It seems to me that it can lead to `use-after-free`

Reply:
Also in usual `Protobuf` format we select specific message name in the schema using setting `format_schema_message_name` setting. What message do we choose here?